File: scripts/modem/memshare-probe.c
/*
 * Publish QMI service 52 (memshare) and log whatever the modem sends to it.
 *
 * This is a diagnostic, not an implementation. The question it answers is
 * whether the modem actually asks the application processor for memory. The
 * stock device tree says it should: qcom,memshare declares a client for the
 * modem with qcom,allocate-boot-time and an 8 MB reserved region, downstream
 * answers on QMI service 0x34, and mainline provides nothing at all. A modem
 * blocked on an allocation that never arrives fits the observed symptoms, which
 * are that it answers queries but never finishes initialising and refuses every
 * operating mode transition.
 *
 * Why C rather than the Python version next to this file: publishing a service
 * by hand over a raw AF_QIPCRTR socket did not work. The service never appeared
 * in qrtr-lookup, which is a silent failure that looks exactly like the modem
 * ignoring you. libqrtr's qrtr_publish() is the path that demonstrably works on
 * this device, since rmtfs, tqftpserv and pd-mapper all register through it.
 *
 * Build on the phone:
 *     sudo apt install libqrtr-dev
 *     gcc -O2 -Wall -o memshare-probe memshare-probe.c -lqrtr
 *
 * Verify it actually registered before trusting any result:
 *     qrtr-lookup | awk '$1==52'
 */

#include <errno.h>
#include <signal.h>
#include <stdbool.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>

#include <libqrtr.h>

#define MEMSHARE_SERVICE  52
#define MEMSHARE_VERSION  1
#define MEMSHARE_INSTANCE 0

static volatile sig_atomic_t stop;

static void on_signal(int signo)
{
	(void)signo;
	stop = 1;
}

static const char *ctrl_name(unsigned int cmd)
{
	switch (cmd) {
	case QRTR_TYPE_HELLO:      return "HELLO";
	case QRTR_TYPE_BYE:        return "BYE";
	case QRTR_TYPE_NEW_SERVER: return "NEW_SERVER";
	case QRTR_TYPE_DEL_SERVER: return "DEL_SERVER";
	case QRTR_TYPE_DEL_CLIENT: return "DEL_CLIENT";
	case QRTR_TYPE_RESUME_TX:  return "RESUME_TX";
	case QRTR_TYPE_EXIT:       return "EXIT";
	case QRTR_TYPE_PING:       return "PING";
	case QRTR_TYPE_NEW_LOOKUP: return "NEW_LOOKUP";
	case QRTR_TYPE_DEL_LOOKUP: return "DEL_LOOKUP";
	default:                   return "unknown";
	}
}

static void hexdump(const unsigned char *buf, unsigned int len)
{
	unsigned int i;

	for (i = 0; i < len; i++) {
		if (i && i % 16 == 0)
			printf("\n        ");
		printf("%02x", buf[i]);
	}
	printf("\n");
}

int main(int argc, char **argv)
{
	unsigned int seconds = argc > 1 ? (unsigned int)atoi(argv[1]) : 120;
	unsigned char buf[4096];
	time_t deadline;
	uint32_t node;
	uint32_t port;
	unsigned int packets = 0;
	int sock;
	int ret;

	signal(SIGINT, on_signal);
	signal(SIGTERM, on_signal);

	/* 0 means "let the kernel pick a port", the same as every other service
	 * on this system that is not rmtfs. */
	sock = qrtr_open(0);
	if (sock < 0) {
		fprintf(stderr, "failed to open qrtr socket: %s\n", strerror(-sock));
		return 1;
	}

	ret = qrtr_publish(sock, MEMSHARE_SERVICE, MEMSHARE_VERSION, MEMSHARE_INSTANCE);
	if (ret < 0) {
		fprintf(stderr, "failed to publish service %d: %s\n",
			MEMSHARE_SERVICE, strerror(-ret));
		return 1;
	}

	printf("published QMI service %d (memshare) v%d instance %d\n",
	       MEMSHARE_SERVICE, MEMSHARE_VERSION, MEMSHARE_INSTANCE);
	printf("check it is real with: qrtr-lookup | awk '$1==%d'\n", MEMSHARE_SERVICE);
	fflush(stdout);

	deadline = time(NULL) + seconds;

	while (!stop && time(NULL) < deadline) {
		ret = qrtr_poll(sock, 1000);
		if (ret < 0) {
			if (errno == EINTR)
				continue;
			fprintf(stderr, "poll failed: %s\n", strerror(errno));
			break;
		}
		if (ret == 0)
			continue;

		ret = qrtr_recvfrom(sock, buf, sizeof(buf), &node, &port);
		if (ret < 0) {
			if (errno == EINTR)
				continue;
			fprintf(stderr, "recv failed: %s\n", strerror(errno));
			break;
		}

		packets++;
		printf("\n--- packet %u from node=%u port=%u, %d bytes\n",
		       packets, node, port, ret);

		if (port == QRTR_PORT_CTRL && ret >= 4) {
			uint32_t cmd;

			memcpy(&cmd, buf, sizeof(cmd));
			printf("    control: %s\n", ctrl_name(cmd));
		} else if (ret >= 7) {
			uint16_t txn;
			uint16_t msg_id;
			uint16_t msg_len;

			memcpy(&txn, buf + 1, sizeof(txn));
			memcpy(&msg_id, buf + 3, sizeof(msg_id));
			memcpy(&msg_len, buf + 5, sizeof(msg_len));
			printf("    QMI type=0x%02x txn=%u msg_id=0x%04x len=%u\n",
			       buf[0], txn, msg_id, msg_len);
			printf("    *** the modem is talking to memshare ***\n");
		}

		printf("        ");
		hexdump(buf, ret > 64 ? 64 : (unsigned int)ret);
		fflush(stdout);
	}

	printf("\ndone: %u packets\n", packets);
	if (!packets)
		printf("nothing arrived. Only meaningful if qrtr-lookup showed"
		       " service %d while this was running\n", MEMSHARE_SERVICE);

	qrtr_close(sock);
	return 0;
}

File: scripts/modem/memshare-probe.py
#!/usr/bin/python3
"""Publish QMI service 52 (memshare) and log whatever the modem sends to it.

This is a diagnostic, not an implementation. The question it answers is whether
the modem actually asks the application processor for memory, which the stock
device tree implies it does (qcom,memshare with a qcom,allocate-boot-time client
for the modem) and which nothing in mainline provides.

If the modem is stalled waiting for that allocation, publishing the service
should make it talk: QRTR notifies waiting clients when a new server appears.

STATUS: **this does not work yet, and its result so far means nothing.** The
script reports that it published, and the modem stays silent, but service 52
never actually appears in `qrtr-lookup` while it runs, whereas rmtfs (14) does.
So the announcement is being dropped somewhere and the modem was never offered
anything to talk to. Do not read the silence as evidence about memshare.

Binding to a fixed port before announcing (below) was one cause of exactly this
and is fixed; something else is still wrong. The obvious next move is to stop
hand rolling the control protocol and use libqrtr's qrtr_publish() from a small
C program, since that is known to work for rmtfs, tqftpserv and pd-mapper on
this very device, and only come back to raw sockets if the C version registers
correctly and more control is needed.
"""

import binascii
import socket
import struct
import sys
import time

QRTR_PORT_CTRL = 0xFFFFFFFE
QRTR_TYPE_NEW_SERVER = 2

MEMSHARE_SERVICE = 52
MEMSHARE_VERSION = 1
MEMSHARE_INSTANCE = 0

# Arbitrary, just needs to be free and non-zero.
LOCAL_PORT = 5052

CTRL_NAMES = {
    0: "HELLO", 1: "BYE", 2: "NEW_SERVER", 3: "DEL_SERVER",
    4: "DEL_CLIENT", 5: "RESUME_TX", 6: "EXIT", 7: "PING",
    8: "NEW_LOOKUP", 9: "DEL_LOOKUP",
}


def publish(sock, node, service, version, instance):
    """Announce a server on this port, the same way libqrtr's qrtr_publish does."""
    pkt = struct.pack("<IIIII",
                      QRTR_TYPE_NEW_SERVER,
                      service,
                      (instance << 8) | version,
                      0,   # node, filled in by the kernel
                      0)   # port, filled in by the kernel
    sock.sendto(pkt, (node, QRTR_PORT_CTRL))


def describe(data):
    """Best effort decode of a QMI header, enough to see what is being asked."""
    if len(data) < 7:
        return "(too short for a QMI header)"
    flags = data[0]
    txn, msg_id, msg_len = struct.unpack_from("<HHH", data, 1)
    kind = {0: "REQUEST", 2: "RESPONSE", 4: "INDICATION"}.get(flags, "type 0x%02x" % flags)
    return "QMI %s txn=%d msg_id=0x%04x len=%d" % (kind, txn, msg_id, msg_len)


def main():
    seconds = int(sys.argv[1]) if len(sys.argv) > 1 else 60

    sock = socket.socket(socket.AF_QIPCRTR, socket.SOCK_DGRAM)
    node, _ = sock.getsockname()

    # Bind to a fixed port before announcing anything. This is not optional:
    # the kernel fills the node and port of a NEW_SERVER message from the
    # sending socket's own address, so publishing from a socket that has no
    # port yet registers a server on port 0. The call appears to succeed and
    # the service simply never shows up in qrtr-lookup, which is a good way to
    # convince yourself the modem ignored you when in fact you never published.
    # Binding to port 0 is rejected outright, so pick one.
    sock.bind((node, LOCAL_PORT))
    node, port = sock.getsockname()
    print("listening as node=%d port=%d" % (node, port), flush=True)

    publish(sock, node, MEMSHARE_SERVICE, MEMSHARE_VERSION, MEMSHARE_INSTANCE)
    print("published QMI service %d (memshare) v%d instance %d"
          % (MEMSHARE_SERVICE, MEMSHARE_VERSION, MEMSHARE_INSTANCE), flush=True)

    sock.settimeout(1.0)
    deadline = time.time() + seconds
    packets = 0

    while time.time() < deadline:
        try:
            data, addr = sock.recvfrom(65536)
        except socket.timeout:
            continue

        packets += 1
        src_node, src_port = addr
        print("\n--- packet %d from node=%d port=%d, %d bytes"
              % (packets, src_node, src_port, len(data)), flush=True)

        if src_port == QRTR_PORT_CTRL and len(data) >= 4:
            cmd = struct.unpack_from("<I", data, 0)[0]
            print("    control: %s" % CTRL_NAMES.get(cmd, "cmd %d" % cmd), flush=True)
        else:
            print("    %s" % describe(data), flush=True)

        print("    %s" % binascii.hexlify(data[:64]).decode(), flush=True)

    print("\ndone: %d packets in %ds" % (packets, seconds), flush=True)
    if not packets:
        print("nothing arrived; the modem did not take up the service", flush=True)


if __name__ == "__main__":
    main()
